File: src/com/aopphp/go/pointcut/NotPointcut.java
package com.aopphp.go.pointcut;

import com.jetbrains.php.lang.psi.elements.PhpClass;
import com.jetbrains.php.lang.psi.elements.PhpClassMember;
import com.jetbrains.php.lang.psi.elements.PhpNamedElement;

import java.util.HashSet;
import java.util.Set;

/**
 * Logic "NOT" pointcut implementation
 */
public class NotPointcut implements Pointcut {

    /**
     * Stores pointcut kind
     */
    protected Set<KindFilter> kind;

    /**
     * Filter for class
     */
    protected PointFilter classFilter = TruePointFilter.getInstance();

    protected Pointcut pointcut;

    public NotPointcut(Pointcut pointcut) {

        this.pointcut = pointcut;
        this.kind     = pointcut.getKind();
    }

    @Override
    public Set<KindFilter> getKind() {
        return kind;
    }

    @Override
    public PointFilter getClassFilter() {
        return classFilter;
    }

    public void setClassFilter(PointFilter classFilter) {
        this.classFilter = classFilter;
    }

    @Override
    public boolean matches(PhpNamedElement element) {
        if (!(element instanceof PhpClassMember)) {
            return false;
        }
        PhpClass preFilter = ((PhpClassMember) element).getContainingClass();
        if (!pointcut.getClassFilter().matches(preFilter)) {
            return true;
        }

        return !pointcut.matches(element);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof NotPointcut)) return false;

        NotPointcut that = (NotPointcut) o;

        if (!kind.equals(that.kind)) return false;
        if (!classFilter.equals(that.classFilter)) return false;
        return pointcut.equals(that.pointcut);
    }

    @Override
    public int hashCode() {
        int result = kind.hashCode();
        result = 31 * result + classFilter.hashCode();
        result = 31 * result + pointcut.hashCode();
        return result;
    }
}

File: src/main/kotlin/com/aopphp/go/pointcut/NotPointcut.kt
package com.aopphp.go.pointcut

import com.jetbrains.php.lang.psi.elements.PhpClassMember
import com.jetbrains.php.lang.psi.elements.PhpNamedElement

class NotPointcut(private val pointcut: Pointcut) : Pointcut {
    private val kind = pointcut.getKind()
    private val classFilter: PointFilter = TruePointFilter

    override fun getKind() = kind
    override fun getClassFilter() = classFilter

    override fun matches(element: PhpNamedElement): Boolean {
        if (element !is PhpClassMember) return false
        val containing = element.containingClass ?: return false
        if (!pointcut.getClassFilter().matches(containing)) return true
        return !pointcut.matches(element)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is NotPointcut) return false
        return kind == other.kind && classFilter == other.classFilter && pointcut == other.pointcut
    }

    override fun hashCode() = 31 * (31 * kind.hashCode() + classFilter.hashCode()) + pointcut.hashCode()
}
